### ISSUE TITLE
Facade Pattern and The I in SOLID

### ISSUE THREAD

**Issue by jackedelic** [nus-cs2103-AY1819S1/forum#211]
In facade pattern, a facade component is introduced to prevent internal details from being exposed to the clients.
On the other hand, according to the Interface Segregation Principle, an interface (abstraction) is introduced so as to prevent a client from being forced to use methods it does not use.
Since software design **patterns** are more concrete than software **principles**, can I treat facade pattern as a direct application of ISP?

Thanks for reading and I appreciate any inputs.

**Comment by anubh-v** [comment]
My understanding is that the Facade pattern and ISP are addressing two separate issues.

Facade pattern provides a solution to the issue where a client does not want to have extensive awareness about the internal pieces of some component - this is solved via the intermediary "facade".

ISP is solving the problems that can occur if a class has to implement useless / redundant behaviour, because it implements an interface that is too broad. This is solved by breaking up the single interface into more modular interfaces - such that each smaller interface contains only related behaviour. 

That's what I understand from some online reading - one article is [here](https://medium.com/@severinperez/avoiding-interface-pollution-with-the-interface-segregation-principle-5d3859c21013)

**Comment by damithc** [comment]
@anubh-v is correct

**Comment by jackedelic** [comment]
Thanks for the clarification (and the online article)! Correct me if I miss out anything, in essence a facade hides internal details (may be useful) from a client, ISP aims to hide non-essential methods/services from a client.

**Comment by jhengy** [comment]
@damithc I understand that facade pattern has  something to do with hiding internal details of a component from the client. Is it only meant to describe the relation between different components at a higher level? Is it still relevant to describe relation between classes? More specifically, given the scenario below, is it an example of facade pattern being demonstrated? (I can very well argue that the internal details of `AdminStaff` are hidden from `Payroll` via the use of the interface `SalariedStaff`(acts as facade))

![image](https://user-images.githubusercontent.com/35480886/49382307-12864080-f751-11e8-90a6-6dfbe960f808.png)

**Comment by damithc** [comment]
@jhengy no, I don't think this case is an example of the Facade pattern.

**Comment by anubh-v** [comment]
@iamjackslayer I think ISP is not about hiding anything, it seems to be about creating appropriate interfaces (a more "cohesive" interface is preferred over one that is too broad)

**Comment by jackedelic** [comment]
@anubh-v I see, much thanks!
